wpengine/wp-graphql-content-blocks
# Implements support for block type filter

This PR creates additional schema changes based on the result of the `get_allowed_block_types` filter.

More specifically:

* If the result of the `get_allowed_block_types` filter returns a list of block names, then we register specialized interfaces for each block type.
For example if `get_allowed_block_types` returns  `['core/paragraph', 'core/image']` for the `post` type then it registers the following schemas:

```
interface PostBlock implements EditorBlock {};

Interface NodeWithPostBlocks implements NodeWithEditorBlocks {
  editorBlocks: [PostBlock];
}
```
* Refactors the code to make the logic in the Registry class more slimmed down.
* Filters out the list of resolved blocks from the response based on the `get_allowed_block_types`. For example if the post content will contain existing blocks not from the allow list they will be filtered out from the GraphQL query.

## How to Test

* Check that by default if you do not provide a filter `get_allowed_block_types` then no schema changes are observed. So there should be no `NodeWithPostBlocks` or `PostBlock` in the schema for example.

* Provide a custom `get_allowed_block_types` filter. Here is an example:
```
add_filter( 'allowed_block_types_all', 'set_allowed_block_types', 10, 2 );

/**
 * Set allowed gutenburg block types
 * https://rudrastyh.com/gutenberg/remove-default-blocks.html
 */
function set_allowed_block_types( $allowed_blocks, $editor_context ) {

	if ( isset( $editor_context->post ) && $editor_context->post instanceof WP_Post && 'post' === $editor_context->post->post_type ) {
		return  [
		   'core/image',
		   'core/paragraph'
		];
	 }
	}
	return true;
}
```

With that filter in place you should be able to query the following fields:

* editorBlocks field for posts should be of type `[PostBlock]`
<img width="1485" alt="Screenshot 2023-04-03 at 13 57 35" src="https://user-images.githubusercontent.com/328805/229516662-21eaecc0-c17a-4833-8310-eaaf2694273c.png">

* editorBlocks field for page should be of type `[EditorBlock]`
<img width="1534" alt="Screenshot 2023-04-03 at 13 59 12" src="https://user-images.githubusercontent.com/328805/229516903-ae8f7bb2-0cce-4b98-ad01-6e941e678fe6.png">

* `PostBlock` implementation should only include the `CoreImage` and `CoreParagraph` blocks.
<img width="301" alt="Screenshot 2023-04-03 at 14 00 48" src="https://user-images.githubusercontent.com/328805/229517309-c8f422a1-d977-42b6-9d14-fac2125921e3.png">

* `EditorBlock` implementation should only include all the available blocks.

* There is a new interface `NodeWithPostBlocks` containing a list of `PostBlock` types:
<img width="1546" alt="Screenshot 2023-04-03 at 14 03 13" src="https://user-images.githubusercontent.com/328805/229517919-5d2d81c1-1008-4b15-a740-5bd36a1e53da.png">

* The list of resolved block data only includes blocks from the allow list. To test that you need to remove the `allowed_block_types_all` filter, create a post with a block not in the previous list  and then enable the filter again. You should be able to see only the filtered blocks:

**Before:**
<img width="1551" alt="Screenshot 2023-04-03 at 14 08 22" src="https://user-images.githubusercontent.com/328805/229519269-c898c7bc-e5ab-42e1-b09b-dd63c643314e.png">

**After:**
<img width="1566" alt="Screenshot 2023-04-03 at 14 09 08" src="https://user-images.githubusercontent.com/328805/229519299-74700b0e-feee-4f6a-9239-cca51022fec8.png">

**Note**. The reason we filter out the blocks from the response is that the resolveType will fail at runtime if we provide a block not implementing the `PostBlock` interface.
After conversation with **jasonbahl** he suggested that there shouldn't be any `UnknownBlock` in case the content resolves a block not in the allow list but instead removed completely.  The caveat here is that it could create some layout blocks not to appear, for example: User has `CoreColumns` in posts. User removes `CoreColumns` from the allowed block list. Then the inner blocks will be orphaned since there is no parent block to host the child blocks.

## Screen Shots
